relay/implementations/beacon-api.ts

```diff
@@ -0,0 +1,323 @@
+    ).json();
+
+    return Number(currentHead.data.header.message.slot);
+  }
+
+  async getBlockSlot(blockHash: string): Promise<number> {
+    const headResult = await (
+      await fetch(`${this.beaconRestApi}/eth/v1/beacon/headers/${blockHash}`)
+    ).json();
+
+    return Number(headResult.data.header.message.slot);
+  }
+
+  async getExistingBlockHeader(slot: number): Promise<
+    ValueOfFields<{
+      slot: UintNumberType;
+      proposerIndex: UintNumberType;
+      parentRoot: ByteVectorType;
+      stateRoot: ByteVectorType;
+      bodyRoot: ByteVectorType;
+    }>
+  > {
+    const { ssz } = await import('@lodestar/types');
+
+    const headResult = await (
+      await fetch(`${this.beaconRestApi}/eth/v1/beacon/headers/${slot}`)
+    ).json();
+
+    return ssz.phase0.BeaconBlockHeader.fromJson(
+      headResult.data.header.message,
+    );
+  }
+
+  async getBlockHeaderOrClosestExisting(
+    slot: number,
+    limitSlot: number,
+  ): Promise<
+    ValueOfFields<{
+      slot: UintNumberType;
+      proposerIndex: UintNumberType;
+      parentRoot: ByteVectorType;
+      stateRoot: ByteVectorType;
+      bodyRoot: ByteVectorType;
+    }>
+  > {
+    const { ssz } = await import('@lodestar/types');
+
+    let blockHeaderResult;
+
+    while (slot <= limitSlot) {
+      blockHeaderResult = await (
+        await fetch(`${this.beaconRestApi}/eth/v1/beacon/headers/${slot}`)
+      ).json();
+
+      if (blockHeaderResult.code !== 404) {
+        return ssz.phase0.BeaconBlockHeader.fromJson(
+          blockHeaderResult.data.header.message,
+        );
+      }
+
+      slot++;
+    }
+
+    throw new Error(
+      `Closest existing block is beyond the limit of ${limitSlot}`,
+    );
+  }
+
+  async getBlockSyncAggregateOrClosestExisting(
+    slot: number,
+    limitSlot: number,
+  ): Promise<{ sync_aggregate: SyncAggregate; slot: number }> {
+    let blockHeaderBodyResult;
+
+    while (slot <= limitSlot) {
+      blockHeaderBodyResult = await (
+        await fetch(`${this.beaconRestApi}/eth/v2/beacon/blocks/${slot}`)
+      ).json();
+
+      if (blockHeaderBodyResult.code !== 404) {
+        return {
+          sync_aggregate:
+            blockHeaderBodyResult.data.message.body.sync_aggregate,
+          slot: slot,
+        };
+      }
+
+      slot++;
+    }
+
+    throw new Error(
+      `Closest existing block is beyond the limit of ${limitSlot}`,
+    );
+  }
+
+  async getPrevBlockHeaderStateInfo(
+    prevSlot: number,
+    nextSlot: number,
+  ): Promise<{
+    finalityHeader: ValueOfFields<{
+      slot: UintNumberType;
+      proposerIndex: UintNumberType;
+      parentRoot: ByteVectorType;
+      stateRoot: ByteVectorType;
+      bodyRoot: ByteVectorType;
+    }>;
+    finalityHeaderBranch: string[];
+    syncCommittee: SyncCommittee;
+    syncCommitteeBranch: string[];
+  }> {
+    const { ssz } = await import('@lodestar/types');
+
+    const prevBeaconStateSZZ = await fetch(
```

Comment:
You can define a helper function for fetching a SSZ state. You seem to have used the same code in few places already.